src/wallet/db.cpp

```diff
@@ -739,6 +747,10 @@ std::string BerkeleyDatabaseVersion()
 
 void BerkeleyDatabase::Release()
 {
+    if (m_active_txn) {
+       m_active_txn->abort();
```

Comment:
In commit "walletdb: Move Txn* functions into WalletDatabase" ([e1ebc69](https://github.com/bitcoin/bitcoin/commit/e1ebc69a5e6628cadf480dfd6d787a6c9f631b86))

re: https://github.com/bitcoin/bitcoin/pull/18971#discussion_r437330343

> It's not coherent to have a shared reference count but only allow a single cursor and a single transaction at a time, and have different batches abort each others transactions and step on each others cursors.

I should be more specific about how I think this could be resolved. I think there are two ways:

1. Add a bunch of asserts to ensure the new shared cursors and transactions aren't misused. `assert(!m_active_txn)` in TxnBegin and Close and Release, `assert(m_active_txn)` in TxnCommit, TxnAbort, Read, Write, etc. `assert(!m_cursor)` in CreateCursor, etc.

2. Don't add shared cursors and transactions. Keep the methods where they are currently in `BerkeleyBatch` instead of moving them to `WalletDatabase`. Just make them them virtual overrides and have `BerkeleyBatch` inherit from a `DatabaseBatch` class. Give `WalletDatabase` a virtual `MakeBatch` method returning `unique_ptr<DatabaseBatch>` and have `WalletBatch` call it during construction and use the pointer as needed. You have two classes with virtual methods instead of one class with virtual methods, but less code moves around and the classes have clear responsibilities.

Another variants of the second approach are also possible like giving WalletBatch itself virtual methods, but the basic idea's the same.

Reply:
Transactions are only used during encryption. Cursors are only used by loading, rewriting, and zapwallettx.

Reply:
I've decided to go with option 1 and added asserts to `CreateCursor` and `TxnBegin`.

I don't think it makes much sense to support multiple transactions and cursors because of our extremely limited usage of them. Transactions are only used when encrypting a wallet. Cursors are used in scenarios where there is only a single thread accessing the database already. These are in loading (both load and zapwallettx) and rewrite, which is encryption. So there won't be concurrent cursors nor concurrent transactions. Even with multiple parallel RPCs, we use locking during encryption to enforce single access, and during loading, the wallet isn't setup yet to even have multiple accesses.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/18971#discussion_r437707266

> Additionally, in SQLite, there aren't such a thing as multiple transactions, so having a separate `Batch/ class would make that more difficult.

RIGHT. :sweat_smile: Finally I understand why you are doing this. I thought you had just gone insane wanting to cram the database and batch classes together and make all these extra changes. But I remember using sqlite in python it doesn't support more than one transaction per connection.

Even so, I don't think this is the best choice or even the easier choice, because it is pushing a problem from the database layer up to the application and RPC layers. Now if there are two RPC calls happening in different threads that both want to open a transaction or a cursor, they have to coordinate with each other to see who is allowed to go first, or risk crashing the process.

Also, the C++ code is more complicated than it needs to be because instead of cursors and transactions being owned by batches and getting properly cleaned up when batches are cleaned up, now they can outlive batches, and error handling and cleanup that should be triggered when the batch goes out of scope now gets lost or delayed until a later point, making debugging more difficult. Also the nice layer separation where Batch objects are responsible for data access, and Database objects are responsible for management is lost, and more code has to move, and everything is more complicated.

If you like the current organization, then please ignore my suggestion, but here is what I would do:

- For BDB, put ownership of transaction and cursor objects back in the batch objects where they are happy and belong.

- For Sqlite, just have `Mutex m_cursor_mutex, m_transaction_mutex` members in the Database instance. Lock m_cursor when a batch object needs a cursor, and unlock when it's done. Similarly for m_transaction. That way if two RPC calls happen, then there aren't errors or aborts, one call just waits for the other.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/18971#discussion_r439666055

> Now if there are two RPC calls happening in different threads that both want to open a transaction or a cursor, they have to coordinate with each other to see who is allowed to go first, or risk crashing the process.

I forgot about cs_wallet when writing this. I think with cs_wallet, two batch objects for the same database will probably not even exist at the same time, so asserts only allowing one cursor or transaction should never trigger, and there should be no need for the `m_transaction_mutex` suggestion. Something like `bool m_transaction_started` and `assert(!m_transaction_started)` for sqlite would be fine. And there wouldn't be a need for `m_cursor_mutex` in any case since there's no problem with multiple cursors, just multiple transactions.

I still think it is better to keep different batch and connection objects for all the other reasons (limiting lifetime of cursor and txn objects, reporting errors earlier, smaller diff, smaller classes, separation of data access methods from database management methods), but with `cs_wallet` the multiple thread RPC request thing should be not be a real issue.

Reply:
There does happen to be a few times where there are multiple `BerkeleyBatch` objects are active, although I don't know where this is happening. You can check this by adding some print lines at the place where `mapFileUseCount` is being updated.

I will probably rework this PR to keep `BerkeleyBatch` and make the move commits at the beginning so this isn't so difficult to update.